### DISCUSSION TITLE
Shared resource in multithreading environment

### DISCUSSION THREAD

**Discussion by TonyFibracci** [uNetworking/uWebSockets#1214] [Q&A] [answered]
let's say i'm running an REST-event-loop on the main thread. It has a PUT that updates a map on the same thread. So far everything is working fine. Now what happens if this map is being accessed by other threads. Will it produce undesirable behavior? 
I know it is not possible to have 2 uWebSocket-event-loops using the same shared resource. In my example i use only 1 but multiple regular threads accessing the same map on the main thread.

**Comment by ghost** [comment, marked as answer]
> I know it is not possible to have 2 uWebSocket-event-loops using the same shared resource.

Lock the map. If it's tiled you only have to lock the tiles accessed. Or shard it. uWS runs on "regular" threads and those can access anything that's properly locked.

**Comment by ibsusu** [comment]
One event loop with multiple threads means any multi-threaded map will work.  You could keep it simple and just lock the map but why do that when you can just link an entire new non-standard library to do it for you? 🤔🤔🤔

Pick your poison:
https://github.com/preshing/junction  a really fast concurrent map library.
https://github.com/oneapi-src/oneTBB intel's thread building blocks.

**Reply by poor1017** [reply]
Hi @ibsusu 

Thank you for your recommendation for junction. I have a problem with using junction. And I can not find much information for this lib.

My situation is here:

1. There is a global hashmap for managing sessions;
2. There are several thread to create session and insert the session to the hashmap from different threads;
3. In each thread, when its work finish, the session is erased from the global hashmap;

Can you tell me:

1. when to call junction::QSBR::Context context = junction::DefaultQSBR.createContext();;
2. when to call junction::DefaultQSBR.update(context);;
3. when to call junction::DefaultQSBR.destroyContext(context);;

Thank you!
